docs/advanced_features.rst

```diff
@@ -0,0 +1,40 @@
+Advanced features
+=================
+
+In this section, you can find some advanced functionalities that are just partially mentioned in the API description.
+
+DALI internal data format
+-------------------------
+
+Internally, DALI processes data in NHWC format. To reshape it to NCHW please use ``\*Permutate`` operator family. It also means that if mentioned operators are used in the middle of the pipeline then next operations will take the wrong assumption about processed data format and the result is undefined.
+
+Thread affinity
+---------------
+
+It is possible to set CPU thread affinity for DALI using ``DALI_AFFINITY_MASK`` environment variable. ``DALI_AFFINITY_MASK`` should be a comma separated list of CPU ids numbers that will be assigned to corresponding DALI threads. A number of DALI threads is set during pipeline construction by ``num_threads`` argument, ``set_affinity`` enables setting thread affinity for the CPU worker threads. It is worth to mention that due to performance reason nvJPEG decoder operator houses threads on its own and they are always affined.
+When ``DALI_AFFINITY_MASK`` is empty affinity is set based on the result returned by nvmlDeviceGetCpuAffinity.
+If there are more threads that values in ``DALI_AFFINITY_MASK`` value from nvmlDeviceGetCpuAffinity will be used for a thread without value provided by an environment variable,
+I.e.:
+
+.. code-block:: bash
+
+  num_threads=5
+  DALI_AFFINITY_MASK=3,5,6,10
+
+Will set thread 0 to CPU 3, 1 to CPU 5, 2 to CPU 6, 3 to CPU 10 and 4 to CPU which is returned by nvmlDeviceGetCpuAffinity.
+
+Memory consumption
+------------------
+
+DALI architecture makes all GPU operators processing the whole batch at the time and CPU one sample at the time. As DALI is only enlarging its memory, never shrinks, each GPU buffer is as large as the largest possible batch, while CPU buffers are as large as batch size multiplied by the size of the largest sample (even though CPU processes one sample at the time per thread still vector of samples need to reside in the memory). This lazy approach helps in cases when memory allocation is expensive like pinned CPU memory or GPU memory reducing the number of total memory operations. This is most visible for the operators which output size may differ from sample to sample, from run to run. Operator with the fixed size outputs - like crop is not affecting overall memory consumption grow over time.
```

Comment:
```suggestion
DALI architecture makes all GPU operators processing the whole batch at the time and CPU one sample at the time. As DALI is only enlarging its memory, never shrinking, each GPU buffer is as large as the largest possible batch, while CPU buffers are as large as batch size multiplied by the size of the largest sample (even though CPU processes one sample at the time per thread still vector of samples need to reside in the memory). This lazy approach helps in cases when memory allocation is expensive like pinned CPU memory or GPU memory reducing the number of total memory operations. This is most visible for the operators which output size may differ from sample to sample, from run to run. Operator with the fixed size outputs - like crop is not affecting overall memory consumption growth over time.
```

Reply:
There is `grow` -> `growth` and `never shrinks` -> `never shrinking`.

Also for CPU, I don't think we resize all the tensors at the same time, only the one that has to fit bigger image.

Reply:
Yes, we don't resize all of them. But during the training the biggest sample can happen at every position in the batch and at the end every CPU would be enlarged to the biggest sample that happened there.